File: admin-api/src/mujahid_admin/services/sessions.py
from __future__ import annotations

from typing import Any

from mujahid_admin.clients.bot import bot_client
from mujahid_admin.core.exceptions import BotNotFoundError
from mujahid_admin.repositories import audit as audit_repo
from mujahid_admin.schemas.session import ActionResponse, QueueTrack, SessionsResponse


async def list_sessions() -> SessionsResponse:
    data = await bot_client.get("/internal/sessions")
    return SessionsResponse.model_validate(data)


async def get_session_queue(guild_id: str) -> list[QueueTrack]:
    try:
        data = await bot_client.get(f"/internal/sessions/{guild_id}/queue")
    except BotNotFoundError as exc:
        raise BotNotFoundError("No active session") from exc
    return [QueueTrack.model_validate(item) for item in data]


async def moderate_session(
    guild_id: str,
    action: str,
    path: str,
    ip: str | None,
) -> ActionResponse:
    result: dict[str, Any] = await bot_client.post(path)
    await audit_repo.write_audit(action, guild_id, ip)
    return ActionResponse.model_validate(result)

File: admin-api/src/services/sessions.ts
import { botClient } from "../clients/bot";
import { BotNotFoundError } from "../core/errors";
import * as auditRepo from "../repositories/audit";
import type { ActionResponse, QueueTrack, SessionsResponse } from "../types";

export async function listSessions(): Promise<SessionsResponse> {
  return botClient.get<SessionsResponse>("/internal/sessions");
}

export async function getSessionQueue(guildId: string): Promise<QueueTrack[]> {
  try {
    return await botClient.get<QueueTrack[]>(`/internal/sessions/${guildId}/queue`);
  } catch (error) {
    if (error instanceof BotNotFoundError) {
      throw new BotNotFoundError("No active session");
    }
    throw error;
  }
}

export async function moderateSession(
  guildId: string,
  action: string,
  path: string,
  ip: string | null,
): Promise<ActionResponse> {
  const result = await botClient.post<ActionResponse>(path);
  await auditRepo.writeAudit(action, guildId, ip);
  return result;
}
